src/odb/src/upf/upf.cpp

```diff
@@ -0,0 +1,160 @@
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include "odb/upf.h"
+
+#include <tcl.h>
+
+#include "utl/Logger.h"
+
+namespace odb {
+
+bool upf::create_power_domain(utl::Logger* logger,
+                              dbBlock* block,
+                              const char* name)
+{
+  return dbPowerDomain::create(block, name) != nullptr;
+}
+
+bool upf::update_power_domain(utl::Logger* logger,
+                              dbBlock* block,
+                              const char* name,
+                              const char* elements)
+{
+  dbPowerDomain* pd = block->findPowerDomain(name);
+  if(pd != nullptr) {
+    pd->addElement(elements);
+  } else {
+    return false;
+  }
+  return true;
+}
+
+bool upf::create_logic_port(utl::Logger* logger,
+                            dbBlock* block,
+                            const char* name,
+                            const char* direction)
+{
+  return dbLogicPort::create(block, name, direction) != nullptr;
+}
+
+bool upf::create_power_switch(utl::Logger* logger,
+                              dbBlock* block,
+                              const char* name,
+                              const char* power_domain,
+                              const char* out_port,
+                              const char* in_port)
+{
+    
+  dbPowerDomain* pd = block->findPowerDomain(power_domain);
+  if(pd == nullptr) return false;
+  dbPowerSwitch* ps = dbPowerSwitch::create(block, name);
+  if (ps == nullptr) return false;
+  ps->setInSupplyPort(in_port);
+  ps->setOutSupplyPort(out_port);
+  ps->setPowerDomain(pd);
+  pd->setPowerSwitch(ps);
+  return true;
+}
+
+bool upf::update_power_switch_control(utl::Logger* logger,
+                                      dbBlock* block,
+                                      const char* name,
+                                      const char* control_port)
+{
+  dbPowerSwitch* ps = block->findPowerSwitch(name);
+  if(ps == nullptr) return false;
+  ps->addControlPort(control_port);
+  return true;
```

Comment:
The return values in update functions are ignored in the .i.  I think it would be better to issue a warning message here than put the handling in the caller.  Same with the other updates